### PR TITLE
[clang][dataflow] Expose fields, globals, and functions referenced.

### PR DESCRIPTION
**Pull request by bazuzi**
Exposes the collection functionality, but does not alter it beyond using a return value instead of output parameters. Also relocates underlying and related functions and a class from DataflowEnvironment's files to DataflowAnalysisContext's files, as no Environment is needed.

### CODE REVIEW

**Comment by bazuzi** [comment]
@ymand Can you review and merge?

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h [hunk 1]
```diff
@@ -62,6 +62,52 @@ FieldSet getObjectFields(QualType Type);
...
+  // member variable because we store a pointer to it in `FieldInits`.
+  std::optional<ImplicitValueInitExpr> ImplicitValueInitForUnion;
+};
+
+struct FieldsGlobalsAndFuncs {
+  FieldSet Fields;
+  // Globals includes all variables with global storage, notably including
+  // static data members and static variables declared within a function.
+  llvm::DenseSet<const VarDecl *> Globals;
+  llvm::DenseSet<const FunctionDecl *> Funcs;
```
Comment (ymand) [inline clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h:106]: Comment `Funcs`? I'm actually not sure offhand what that's for. :)

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h [hunk 2]
```diff
@@ -62,6 +62,52 @@ FieldSet getObjectFields(QualType Type);
...
+private:
+  SmallVector<std::pair<const CXXBaseSpecifier *, Expr *>> BaseInits;
+  SmallVector<std::pair<const FieldDecl *, Expr *>> FieldInits;
+
+  // We potentially synthesize an `ImplicitValueInitExpr` for unions. It's a
+  // member variable because we store a pointer to it in `FieldInits`.
+  std::optional<ImplicitValueInitExpr> ImplicitValueInitForUnion;
+};
+
+struct FieldsGlobalsAndFuncs {
```
Comment (ymand) [inline clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h:101]: Maybe prefix "Mentioned" or "Referenced" or "Used"?
Also, please add a brief comment. In context, it's role is obvious, but for cross-referencing tools, its often helpful to have comment when you're looking at a mention in another file.

**Review by ymand — APPROVED** [review]
Thanks!

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h [hunk 3]
```diff
@@ -62,6 +62,52 @@ FieldSet getObjectFields(QualType Type);
 bool containsSameFields(const FieldSet &Fields,
                         const RecordStorageLocation::FieldToLoc &FieldLocs);
 
+/// Returns the fields of a `RecordDecl` that are initialized by an
+/// `InitListExpr`, in the order in which they appear in
+/// `InitListExpr::inits()`.
+/// `Init->getType()` must be a record type.
+std::vector<const FieldDecl *>
+getFieldsForInitListExpr(const InitListExpr *InitList);
```
Comment (martinboehme) [inline clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h:70]: This appears to be used only in DataflowEnvironment.cpp -- would suggest simply making it a static function there (not exposed in any header).

(I believe it used to be that this was used more widely, but `RecordInitListHelper` made that unnecessary, and I didn't notice that when I introduced this class?)

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h (hunk 2 above)
Comment (martinboehme) [reply clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h:101]: My vote would be for `Referenced`, and I would also suggest replacing `FieldsGlobalsAndFuncs` (which is a name that IIRC has "accreted" over time). How about `ReferencedDecls` (and `getReferencedDecls()`), which reads better and will remain accurate if we add more types of declarations in the future?

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h [hunk 4]
```diff
@@ -62,6 +62,52 @@ FieldSet getObjectFields(QualType Type);
...
+
+struct FieldsGlobalsAndFuncs {
+  FieldSet Fields;
+  // Globals includes all variables with global storage, notably including
+  // static data members and static variables declared within a function.
+  llvm::DenseSet<const VarDecl *> Globals;
+  llvm::DenseSet<const FunctionDecl *> Funcs;
+};
+
+FieldsGlobalsAndFuncs getFieldsGlobalsAndFuncs(const FunctionDecl &FD);
```
Comment (martinboehme) [inline clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h:109]: Can you add a short comment for this function?

**Review by martinboehme — CHANGES_REQUESTED** [review]

**Comment by bazuzi** [comment]
Sorry, screwed up my git push here. Going to close this to avoid hitting recent authors any more.

**Pull request closed without merging**